### PR TITLE
feat: add gridview in backend and android

### PR DESCRIPTION
**Pull request by uziasferreirazup**
### Related Issues

#1640

### Description and Example

Add `direction` and `spanCount` to the GridView. Deprecate `numColumns`. When `numColumns` is passed instead of `direction` and `spanCount`, consider `direction = "vertical"` and `spanCount = numColumns`. If all are passed, ignore `numColumns`.

### Checklist

- [x] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [x] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [x] I am willing to follow-up on review comments in a timely manner.
- [x] I have made the documentation changes or I created an issue explaining how to document this change on [Docs issues](https://github.com/ZupIT/beagle-docs/issues). Please link the issue here:

[PR Guide]: https://github.com/ZupIT/beagle/blob/master/doc/contributing/pull_requests.md

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt [hunk 1]
```diff
@@ -22,8 +22,10 @@ import android.view.View
 import android.view.ViewGroup
 import androidx.fragment.app.Fragment
 import br.com.zup.beagle.android.components.GridView
+import br.com.zup.beagle.android.components.GridViewDirection
 import br.com.zup.beagle.android.components.Image
```
Comment (botbeaglezup) [inline android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt:26]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/d4d30c2c124c791521835c4f431e85247fa3f3c4/android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt#L26"></span>The import 'br.com.zup.beagle.android.components.Image' is unused.</td>
    </tr>
  </tbody>
</table>

File: android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt [hunk 2]
```diff
@@ -22,8 +22,10 @@ import android.view.View
...
 import br.com.zup.beagle.android.components.ImagePath
```
Comment (botbeaglezup) [inline android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt:27]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/d4d30c2c124c791521835c4f431e85247fa3f3c4/android/sample/src/main/java/br/com/zup/beagle/sample/fragment/GridViewFragment.kt#L27"></span>The import 'br.com.zup.beagle.android.components.ImagePath' is unused.</td>
    </tr>
  </tbody>
</table>

**Review by matheusribeirozup — APPROVED** [review]

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>

### Tests: 

Name | Classname | Time|
--- | --- | ---|
Given  a BeagleFragment with screen identifier When BeagleFragment is resumed Then should report screen | br.com.zup.beagle.android.view.BeagleFragmentTest | 4.977|


### Tests: 

Name | Classname | Time|
--- | --- | ---|
Given a ListAdapter with horizontal orientation When call onCreateViewHolder Then should call setWidthAutoAndDirtyAllViews for the created itemView | br.com.zup.beagle.android.components.list.ListAdapterTest | 13.414|
Given a ListAdapter When call onViewRecycled Then should call viewHolder onViewRecycled | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.078|
Given a ListAdapter When a view executes an async action and there is no ViewHolder created for the adapter Then action must not be observed | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.06|
Given a ListAdapter When call onCreateViewHolder Then should create the default template | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.073|
Given a ListAdapter When call getItemViewType Then should get the correct default template index | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.066|
Given a ListAdapter When call setList with a non null list Then adapter item count should return list size | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.053|
Given a ListAdapter When call getItemViewType Then should evaluate context only once for each adapter position | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.045|
Given a ListAdapter with vertical orientation When call onCreateViewHolder Then should call setHeightAutoAndDirtyAllViews for the created itemView | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.045|
Given a ListAdapter When call onBindViewHolder Then should call viewHolder onBind | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.039|
Given a ListAdapter with vertical orientation When call onCreateViewHolder Then should set the correct layout params to view holder itemView | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.042|
Given a ListAdapter When a view holder itemView call an async action and action has status started Then should make holder not recyclable | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.046|
Given a ListAdapter When ListAdapter is initialized Then it should register to observe all async actions executed | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.038|
Given a ListAdapter with horizontal orientationWhen call onCreateViewHolder Then should set the correct layout params to view holder itemView | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.037|
Given a ListAdapter When call setList and adapter recyclerId is -1 Then should generate a temporary id | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.053|
Given a ListAdapter When call setList with a non null list Then should notify ListViewIdViewModel | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.058|
Given a ListAdapter When call setRecyclerId with value -1 Then should not change adapter recyclerId | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.048|
Given a ListAdapter When call onCreateViewHolder Then should create the correct template | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.057|
Given a ListAdapter When a view with a null parent executes an async action Then holder should not observe that action | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.048|
Given a ListAdapter When call onCreateViewHolder Then should create an itemView with autoGenerateIdEnabled false | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.075|
Given a ListAdapter When a view holder itemView call an async action and action has status finished Then should make holder recyclable | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.04|
Given a ListAdapter When a view executes an async action and that view is not related to adapter ViewHolders Then action must not be observed | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.049|
Given a ListAdapter When call getItemViewType after calling setList Then should evaluate context for each adapter position again | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.047|
Given a ListAdapter When call setRecyclerId with value different of -1 Then should change adapter recyclerId | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.041|
Given a ListAdapter When call onCreateViewHolder Then should create the non existent default template | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.036|
Given a ListAdapter with a null template list When call onCreateViewHolder Then should create the template | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.048|
Given a ListAdapter When call clone Then cloned adapter should have exactly the same properties of source adapter | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.048|
Given a ListAdapter When call onViewAttachedToWindow Then should call viewHolder onViewAttachedToWindow | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.041|
Given a ListAdapter When call setList and adapter recyclerId is different of -1 Then should not generate a temporary id | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.036|
Given a ListAdapter When call setList with a null list Then adapter item count should be zero | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.036|
Given a ListAdapter When a view that is not inside a RecyclerView executes an async action Then holder should not observe that action | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.036|
Given a ListAdapter When call getItemViewType Then should return -1 when there is no default template | br.com.zup.beagle.android.components.list.ListAdapterTest | 0.035|


### Tests: 

Name | Classname | Time|
--- | --- | ---|
Given a screen with identifier When navigate to Then should call BeagleFragment newInstance with right parameters | br.com.zup.beagle.android.view.BeagleActivityTest | 0.893|
Given a screen with id When navigate to Then should call BeagleFragment newInstance with right parameters | br.com.zup.beagle.android.view.BeagleActivityTest | 0.259|
Given a screen request When navigate to Then should call BeagleFragment newInstance with right parameters | br.com.zup.beagle.android.view.BeagleActivityTest | 0.292|


### Tests: 

Name | Classname | Time|
--- | --- | ---|
Given a Beagle View When loadView with Request Data Then should call fetch component | br.com.zup.beagle.android.view.custom.BeagleViewTest | 0.9|
Given a DoRenderState with screen id null When loadView Then Should not ReportScreen | br.com.zup.beagle.android.view.custom.BeagleViewTest | 0.563|
Given a DoRenderState with a ScreenIdentifier not null When loadView Then Should ReportScreen | br.com.zup.beagle.android.view.custom.BeagleViewTest | 0.554|


<p align="right" data-meta="generated_by_android">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by matheusribeirozup — APPROVED** [review]

**Review by viniciusguardieirozup — APPROVED** [review]

**Pull request merged**